AdyenDropIn/Modules/PaymentMethodList/PaymentMethodListRouter.swift

```diff
@@ -49,14 +51,51 @@ internal class PaymentMethodListRouter: PaymentMethodListRouterProtocol {
         navigationController.setViewControllers([view], animated: false)
     }
 
-    internal func cancel(completion: (() -> Void)?) {
-        delegate?.cancelPayment(completion: completion)
+    // MARK: - Private
+}
+
+extension PaymentMethodListRouter: PaymentMethodListViewModelDelegate {
+
+    // MARK: - PaymentMethodListViewModelDelegate
+
+    func didCancel(completion: (() -> Void)?) {
+        delegate?.paymentMethodListDidCancel(completion: completion)
+        componentContainerRouter = nil
     }
 
     internal func didSelect(_ component: PresentableComponent) {
-        let componentViewController = componentContainerAssembler.resolveContainerView(for: component)
-        view?.navigationController?.pushViewController(componentViewController, animated: true)
+        let componentContainerRouter = componentContainerAssembler.resolveComponentContainerRouter(
+            for: component,
+            delegate: self
+        )
+        self.componentContainerRouter = componentContainerRouter
+
+        componentContainerRouter.start()
+
+        let componentContainerViewController = componentContainerRouter.rootViewController
+
+        // TODO: - Invert `requiresModalPresentation` logic or remove it fully.
+        if component.requiresModalPresentation {
+            view?.navigationController?.pushViewController(componentContainerViewController, animated: true)
```

Comment:
I think it adds unnecessary UIKit internals exposure to an abstract router. In this case this could be extracted to a `presenter.push()` that encapsulates UIKit presentation logic and maybe later could be hot-swapped with another presenter. It would also make it testable.

Reply:
I want to revisit this, not finished yet. Good idea the presenter, I will consider it.